gpt_engineer/core/chat_to_files.py

```diff
@@ -100,97 +100,88 @@ def overwrite_code_with_edits(chat: str, files_dict: FilesDict):
 @dataclass
 class Edit:
     filename: str
-    before: str
-    after: str
+    line_number: int
+    content: str
+    is_before: bool
 
 
 def parse_edits(chat: str):
     """
-    Parse edits from a chat string.
-
-    This function extracts code edits from a chat string and returns them as a list
-    of Edit objects.
-
-    Parameters
-    ----------
-    chat : str
-        The chat content containing code edits.
-
-    Returns
-    -------
-    List[Edit]
-        A list of Edit objects representing the parsed code edits.
+    Parse edits from a chat string, preserving indentation.
     """
-
-    def parse_one_edit(lines):
-        HEAD = "<<<<<<< HEAD"
-        DIVIDER = "\n=======\n"
-        UPDATE = ">>>>>>> updated"
-
-        filename = lines.pop(0)
-        text = "\n".join(lines)
-        splits = text.split(DIVIDER)
-        if len(splits) != 2:
-            raise ValueError(f"Could not parse following text as code edit: \n{text}")
-        before, after = splits
-
-        before = before.replace(HEAD, "").strip()
-        after = after.replace(UPDATE, "").strip()
-
-        return Edit(filename, before, after)
-
     edits = []
-    current_edit = []
-    in_fence = False
-
-    for line in chat.split("\n"):
-        if line.startswith("```") and in_fence:
-            edits.append(parse_one_edit(current_edit))
-            current_edit = []
-            in_fence = False
-            continue
-        elif line.startswith("```") and not in_fence:
-            in_fence = True
-            continue
-
-        if in_fence:
-            current_edit.append(line)
-
+    lines = chat.split("\n")
+    filename = ""
+    for line in lines:
+        if line.startswith("File:"):
+            filename = line.split(":")[1].strip()
+        else:
+            # Modified regex to capture leading spaces (indentation)
+            match = re.match(r"(\d+) ([-+]) (.*\S.*)", line)
+            if match:
+                line_number, symbol, content = match.groups()
+                line_number = int(line_number)
+                is_before = symbol == "-"
+                edits.append(Edit(filename, line_number, content, is_before))
+
+    # Sort edits such that 'before' edits come before 'after' edits
+    edits.sort(key=lambda edit: (edit.filename, edit.line_number, not edit.is_before))
     return edits
 
 
 def apply_edits(edits: List[Edit], files_dict: FilesDict):
-    """
-    Apply a list of edits to the given code.
+    # Helper function to compare strings ignoring spaces
+    def is_similar(str1, str2):
```

Comment:
I think this criterion could be relaxed further. For example, you could count the number each  character occurs (unordered) and then require that the intersection between the code in the file and the edit needs to be 90% the lengths of the line in the file or something like that. For short lines (like 1 character), the criterion should require a perfect match. This criterion should be called both by apply edits and in the incorrect_edit function that I referenced in the other comment.